optuna/notebooks/jupytercon_demo/RAPIDS_xfeat_Optuna-GPU.ipynb

```diff
@@ -0,0 +1,1803 @@
+    "            \n",
+    "            mlflow.log_params(trial.params)\n",
+    "\n",
+    "            tags = {}\n",
+    "            tags[\"number\"] = str(trial.number)\n",
+    "            tags[\"datetime_start\"] = str(trial.datetime_start)\n",
+    "            tags[\"datetime_complete\"] = str(trial.datetime_complete)\n",
+    "            tags['RAPIDS cuDF Version'] = str(cudf.__version__)\n",
+    "            tags['RAPIDS cuML Version'] = str(cuml.__version__)\n",
+    "            tags['SKlearn Version'] = str(sklearn.__version__)\n",
+    "\n",
+    "            trial_state = trial.state\n",
+    "            if (isinstance(trial_state, TrialState)):\n",
+    "                tags['state'] = str(trial_state).split('.')[-1]\n",
+    "            \n",
+    "            # Set direction and convert it to str and remove the common prefix.\n",
+    "            study_direction = study.direction\n",
+    "            if isinstance(study_direction, StudyDirection):\n",
+    "                tags[\"direction\"] = str(study_direction).split(\".\")[-1]\n",
+    "\n",
+    "            tags.update(trial.user_attrs)\n",
+    "            distributions = {\n",
+    "                (k + \"_distribution\"): str(v) for (k, v) in trial.distributions.items()\n",
+    "            }\n",
+    "            tags.update(distributions)\n",
+    "\n",
+    "            # This is a temporary fix on Optuna side. It avoids an error with user\n",
+    "            # attributes that are too long. It should be fixed on MLflow side later.\n",
+    "            # When it is fixed on MLflow side this codeblock can be removed.\n",
+    "            # see https://github.com/optuna/optuna/issues/1340\n",
+    "            # see https://github.com/mlflow/mlflow/issues/2931\n",
+    "            max_mlflow_tag_length = 5000\n",
+    "            for key, value in tags.items():\n",
+    "                value = str(value)  # make sure it is a string\n",
+    "                if len(value) > max_mlflow_tag_length:\n",
+    "                    tags[key] = textwrap.shorten(value, max_mlflow_tag_length)\n",
+    "\n",
+    "            mlflow.set_tags(tags) "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Feature Engineering\n",
+    "\n",
+    "The following functions are defined to perform a few feature engineering tasks on the data. The feature_engineering function is called on the dataframe df, in this function we use cuML's [LabelEncoder](https://docs.rapids.ai/api/cuml/stable/api.html#cuml.preprocessing.LabelEncoder.LabelEncoder) and [TargetEncoder](https://docs.rapids.ai/api/cuml/stable/api.html#cuml.preprocessing.TargetEncoder) to encode the categorical columns. Arithmetic Combinations on the numerical columns using xfeat. The Arithmetic Operation performed here is addition on two columns at once. We specify the operator and r, which is used to indicate how many columns need to be combined.\n",
```

Comment:
You explain this more later, but maybe mention explicitly here something like "We're just going to create a LOT of features without much manual optimization. This is part of the charm of automated feature engineering - the next steps will help make sure that only the most useful generated features are kept."